rfcs/0125-bootspec.md

```diff
@@ -0,0 +1,216 @@
+  Concretely, making it relatively straightforward to convert most of the empty spaces in the feature matrix to YES's.
+- Enable users of NixOS to implement custom bootloader tools and policy without needing to dive through the system profiles, and without patching Nixpkgs / NixOS.
+- Define a stable specification of a generation's boot data which internal and external users can rely on.
+  Changes to the specification should go through an RFC.
+
+### Non-Goals
+[non-goals]: #non-goals
+
+- Rewriting the existing bootloader backends to actually fill out the feature matrix.
+  The goal of this RFC is to make the feature development *easier*, not actually do it.
+- Supporting SecureBoot.
+  The authors of this RFC have done work in this regard, but this RFC is not about SecureBoot.
+- Integrating Bootspec into an existing bootloader backend is not expected to perfectly produce the exact same result. For example, it is likely that the text describing the entrties in the menu may differ.
+
+# Proposed Solution
+[proposed-solution]: #proposed-solution
+
+- Each NixOS generation will have a bootspec (a JSON document) at `$out/bootspec/boot.v1.json` containing all of the boot properties for that generation.
+  NixOS's bootloader backends will read these files as inputs to the bootloader installation phase.
+  A bootspec document named `boot.v1.json` must have the `schemaVersion` set to `1`. Any other value makes the document invalid. Generally, the file must be named `boot.v${schemaVersion}.json`.
+    - The bootloader installation phase is relatively unchanged from the way it is now.
+      The bootloader backend will have an executable that is run against a collection of generations, and the backend is any of the currently supported backends plus an "external" backend which the user can define.
+- The bootloader backends will avoid reading data from the other files and directories when possible, preferring the information in the bootspec.
+- A bootspec synthesizing tool will be used to synthesize a bootspec for generations which don't already have one.
+  This tool will be shared across all of the bootloader backends, helping produce more uniform behavior.
+- Existing bootloader backends will be updated to read properties from the bootspec, removing most if not all of their filesystem-spelunking code.
+
+### Bootspec Format v1
+[format-v1]: #format-v1
+
+
+Using the following JSON:
+
+```json5
+{
+  // Version of the specification used in the document
+  "schemaVersion": 1,
+
+  // Path to the stage-2 init, executed by the initrd
+  "init": "/nix/store/xxx-nixos-system-xxx/init",
+
+  // Path to the initrd
+  "initrd": "/nix/store/xxx-initrd-linux/initrd",
+
+  // Optional path to a tool to dynamically add secrets to an initrd
+  "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",
+
+  // Path to the kernel image
+  "kernel": "/nix/store/xxx-linux/bzImage",
+
+  // Kernel commandline options
+  "kernelParams": [
```

Comment:
But kernel command line is also used to pass configuration to early init, so it is a bit less limited.

A list gives an option to do at least some filtering.

Reply:
There's no drawbacks to keeping this a list, while if it was squished in a string it would lose a bit of the meaning of the input. At least that's what I see here.

> A list gives an option to do at least some filtering.

It's probably too late to filter at this point; this is an input for tools that serialize the generation's boot configuration. But true, it wouldn't be as easy to filter otherwise.

Reply:
I assume that a) there will be unpredictable mix of data for kernel / PID1 / activation, b) the type choice here will propagate into some random place.

So a list costs little, allows asserting lack of spaces in entries if we want (pretty sure _something_ will be broken by them), and increases chances to filter out a single specific option before bootspec being written. Overall, a win, in my opinion.